[Jonathan Tay] iP

jontmy opened the pull request:
### Changes

> v0.3.0

- Added a GUI to Duke wrapping around the CLI

### Tasks

- [ ] Refactor the UI code for proper OOP


### Miscellany

1. Here is a snippet of a [test](https://github.com/jontmy/ip/blob/master/src/test/java/duke/task/TodoTest.java) using `JUnit5`:
```java
public class TodoTest {
    @Test
    public void testToString() {
        assertEquals("[T][ ] abc", new Todo("abc").toString());
    }
}
```

**HAND** 👋

jontmy committed:
Implement command echo and exit

jontmy committed:
Create class `TaskList`

jontmy committed:
Implement text storing

jontmy committed:
Fix erroneous `static` declaration in `TaskList`

jontmy committed:
Create class `Task`

jontmy committed:
Refactor `TaskList` to use `Task`

jontmy committed:
Implement level 3 requirements

jontmy committed:
Add responses to some commands

jontmy committed:
Add check for empty commands

jontmy committed:
Convert `Task` into an abstract class

jontmy committed:
Create the 3 subclasses of `Task`

jontmy committed:
Add method to get the size of a `TaskList`

jontmy committed:
Implement level 4 requirements

jontmy committed:
Implement level 5 requirements

jontmy committed:
Fix todo command

jontmy committed:
Fix newline insertions

jontmy committed:
Fix newline insertions

jontmy committed:
Add editorconfig

jontmy committed:
Implement automated test requirements

jontmy committed:
Implement level 6 requirements

jontmy committed:
Fix whitespace being ignored in commands

jontmy committed:
Implement task serialization and deserialization

jontmy committed:
Implement rest of level 7 requirements

- Refactored `Duke` to be a singleton (otherwise conflicts may occur in the hardcoded file)
- Implemented task list saving after each command execution

jontmy committed:
Refactor `Task` and its subclasses to use LocalDateTime

jontmy committed:
Merge branch `branch-Level-7` into `master`

jontmy committed:
Merge branch `branch-Level-8` into `master`

jontmy committed:
Fix incorrect access modifier for `Duke` singleton getter

jontmy committed:
Fix typo in documentation

jontmy committed:
Add Gradle

jontmy committed:
Refactor codebase to use packages

jontmy committed:
Change gradle config to use packages

jontmy committed:
Update gradle config for JUnit

jontmy committed:
Write tests for `TaskList` and `Todo`

jontmy committed:
Fix parsing of years in dates

jontmy committed:
Add skeleton classes

jontmy committed:
Create interface `Streamable`

jontmy committed:
Refactor commands

jontmy committed:
Fix date serialized format

jontmy committed:
Implement requirements for A-MoreOOP

jontmy committed:
Merge branch `branch-A-MoreOOP` into `master`

jontmy committed:
Implement requirements for Level-9

jontmy committed:
Replace `*` imports to conform with coding standard

jontmy committed:
De-indent `case` blocks to conform with coding standard

jontmy committed:
Add JavaDocs

jontmy committed:
Merge branch `branch-Level-9` into `master`

jontmy committed:
Merge branch `branch-A-JavaDoc` into `master`

jontmy committed:
Merge branch `A-CodingStandard` into `master`

jontmy committed:
Add JavaFX dependencies

jontmy committed:
Add tutorial code

jontmy committed:
Refactor codebase into separate directories

jontmy committed:
Rename `duke.internals` to `duke.internal`

jontmy committed:
Re-implement CLI features in a GUI

jontmy committed:
Merge branch `branch-Level-10` into `master`

jontmy committed:
Add checkstyle and related config files

jontmy committed:
Fix all non-Javadoc checkstyle warnings

jontmy committed:
Merge branch `branch-A-CheckStyle` into `master`

jontmy committed:
Fix whitespace in JavaDocs

farisanadia commented on src/main/java/duke/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,35 @@
+ * [0]: index of the task to unmark as done
+ * Note that the index starts at 1.
+ */
+public class UnmarkCommand extends Command {
```
maybe this should be unmarkCommand

farisanadia commented on src/main/java/duke/internal/Parser.java:
```diff
@@ -0,0 +1,145 @@
+
+import java.util.Arrays;
+import java.util.stream.IntStream;
+
```
maybe you shouldn't leave a space here for consistency purposes?

farisanadia commented on src/main/java/duke/internal/Parser.java:
```diff
@@ -0,0 +1,145 @@
+     * @param str the string to parse
+     * @return the Command object to execute
+     */
+    public static Command parse(String str) {
```
could this be called parseCommand instead? 
_"Names representing methods must be verbs and written in camelCase."_

farisanadia replied:
could this be called parseCommand instead? 
_"Names representing methods must be verbs and written in camelCase."_

farisanadia commented on src/main/java/duke/internal/Storage.java:
```diff
@@ -0,0 +1,56 @@
+package duke.internal;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import duke.task.Task;
+import duke.task.TaskList;
+
```
similarly, maybe auto ordering can be used here for consistent import statements

farisanadia commented on src/main/java/duke/internal/Storage.java:
```diff
@@ -0,0 +1,56 @@
+     * @throws IOException if an I/O error occurs writing to the file
+     * @since Level-7
+     */
+    public void save(TaskList tasks) throws IOException {
```
could this be called saveTask and loadTask instead?

farisanadia commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,50 @@
+ * A task that start at a specific date/time and ends at a specific date/time.
+ */
+public class Event extends Task {
+    private final LocalDateTime datetime;
```
should this be called dateTime instead?
_" Variable names must be in camelCase."_

farisanadia replied:
should this be called dateTime instead?
_" Variable names must be in camelCase."_

farisanadia commented on src/main/java/duke/internal/Storage.java:
```diff
@@ -0,0 +1,56 @@
+     * @throws IOException if an I/O error occurs writing to the file
+     * @since Level-7
+     */
+    public void save(TaskList tasks) throws IOException {
```
could this be called saveTask and loadTask instead?

jontmy replied:
Good suggestion! I'll amend this in the next iteration.

farisanadia commented on src/main/java/duke/internal/Storage.java:
```diff
@@ -0,0 +1,56 @@
+package duke.internal;
+
+import java.io.IOException;
```
similarly, maybe auto ordering can be used here for consistent import statements

farisanadia commented on src/main/java/duke/internal/Parser.java:
```diff
@@ -0,0 +1,145 @@
+package duke.internal;
+
+import java.util.Arrays;
+import java.util.stream.IntStream;
+
+import duke.command.ByeCommand;
+import duke.command.Command;
+import duke.command.DeadlineCommand;
```
maybe you shouldn't leave a space here for consistency purposes?

jontmy replied:
Hi! Just wanted to point out that the [coding standard](https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements) actually uses newlines to group related packages together.

The AB3 checkstyle also enforces this newline:
 
<img width="871" alt="Screenshot 2022-09-03 at 13 13 05" src="https://user-images.githubusercontent.com/50949210/188256843-951805a2-2f06-4da4-9f08-089f4430032d.png">

farisanadia replied:
ah i see! my bad

farisanadia reviewed (commented):
Overall, your code is easy to understand with good javadocs. Well done!

jontmy committed:
Improve code quality

- `Event.datetime` renamed to `Event.dateTime` for proper camel casing
- `Parser::parse` renamed to `Parser::parseString` for clarity
- `Storage::load` and `Storage::save` renamed to `loadTasks` and `saveTasks`
   respectively

jontmy committed:
Add assertions

- Assert that delimiters are specified in `Parser::parse`
  to ensure proper control flow

jontmy committed:
Merge pull request #1 from jontmy/branch-A-CodeQuality

jontmy committed:
Merge branch `master` into `branch-A-Assertions`

jontmy committed:
Merge pull request #2 from jontmy/branch-A-Assertions

jontmy committed:
Add CI workflow

jontmy committed:
Add default aliases

jontmy committed:
Fix checkstyle configFile setting causing CI failure

jontmy committed:
Merge branch `master` into `branch-C-FriendlierSyntax`

jontmy committed:
Add command aliasing

jontmy committed:
Fix checkstyle configuration

jontmy committed:
Add JavaDocs

jontmy committed:
Fix application build config

jontmy committed:
Add image of the Knight

jontmy committed:
Replace image of Duke with the Knight

jontmy committed:
Add custom font

jontmy committed:
Remove ASCII text art

jontmy committed:
Replace user image with Hornet

jontmy committed:
Delete unused images

jontmy committed:
Modify GUI for better UX

- Messages now wrap instead of being truncated with an ellipsis
- Increased font size to 16pt for readability

jontmy committed:
Update Knight's personality

jontmy committed:
Merge branch `branch-A-Personality` into `master`

jontmy committed:
Add stage icon and title

jontmy committed:
Increase default stage size

jontmy committed:
Refactor GUI classes for better OOP

jontmy committed:
Rename class `ui` to `MessageBuilder`

jontmy committed:
Add image attributions

jontmy committed:
Add set of default commands

jontmy committed:
Refactor for SLAP

jontmy committed:
Merge branch `branch-A-BetterGui` into `master`

jontmy committed:
Replace existing images with copyright-free images

jontmy committed:
Add UI screenshot

jontmy committed:
Create user guide

jontmy committed:
Merge branch `branch-A-UserGuide` into `master`

jontmy committed:
Move command summary to the top

jontmy committed:
Update readme

jontmy committed:
Fix malformed JavaDoc

jontmy committed:
Add checks for internal delimiter usage to prevent file corruption

jontmy committed:
Change `IllegalArgumentException` added in previous commit to `DukeException`
